Macros

Cryptjar opened the pull request:
Implement #7: a formatting macro and example demonstrating the usage.

The precise syntax of the implemented macro is defined as follows:

Variable: _any identifier_
Color: `red` | `green` | `blue` | ...
Style: `bold` | `italic` | `reverse` | `underline`
Simple: Color | `#` Color | Style | `!` Style | `reset` | `!fg` | `!bg` | `!style`
Complex: `fg=` Variable | `bg=` Variable | `style=` Variable | `style+=` Variable | `style-=` Variable | `=` Variable
Format: `[` Simple `]` | `[` Complex `]`

Literal: _any string literal or number_
Expr: _any Rust expression_
RFS: _format string literal followed by any number of expressions as arguments_
Text: Literal | `(` RFS `)` | `(:` Expr `)` | `(?` Expr `)`

MacroInput: ( `lock` )? Variable `;` ( Format | Text )*

murarth commented:
This looks great. Just a few issues:

* Why are `,,;,;;` and `;;;,;,` valid input for the macro?
* It seems unnecessary to shorten `style` to `sty`.
* I don't like the use of `.unwrap()` within the macro, though I understand that it's not exactly straightforward to chain `Result` values because `Screen` methods all return `()`. I have an idea for how to solve this problem. I'll get back to you after I do some testing to see if it works.

Cryptjar commented:
Oh, that thing about the commas and semicolons, I once though it might be a good idea to allow to separate the different tokens by comma or semicolon while keeping it optional. As a result it is possible to also separate empty tokens, by just writing commas and semicolons alone. Since I made the square brackets mandatory in the end, I guess the comma and semicolon stuff became superfluous. So I removed them.

About your second point, well, depending on taste, the `sty` thing might be unnecessary, in deed. However, it seems to be not consistent with the rest of the API, so I removed them too.

About your last point, it is in deed difficult. As I see it, one problem is the incompatibility between the `Screen` API (returns `()`) and the `Terminal` API (returns `Result`). The other problem is when one call the `.unwrap()` in the macro the user can't do any error handling and the program might just crash in case there is an `Err` hiding somewhere, but, to the best of my knowledge, the `macro_rules!` macros do not support expression-style macros meaning that the macro may not 'return' the `Result` but instead may only ignore the `Result`s usually yielding lots of compiler warnings wherever the macro is used.

murarth commented:
I've just pushed a set of commits, based on this PR, to improve your implementation of the `term_write` macro. I've made the following changes:

* Removed the `lock` syntax from the macro; a new method `borrow_term_write_guard` was added to the `Terminal` and `Screen` structs, as well as their respective write guard types. This method is marked `doc(hidden)`, as it is an implementation detail of the macro and not intended for external use.  
For `Terminal` and `Screen` instances, this method acquires the write lock and returns it. For lock guard instances, it simply returns `self`.  
This allows the macro to either acquire a lock and hold it while performing each operation or to use a lock provided without consuming it.
* Added result value chaining, using a trait (which is also marked `doc(hidden)` for the same reason as above) that is implemented for both `()` and `io::Result<()>`. This allows the macro to accept either a `Screen` or `Terminal` instance and, in the latter case, to yield an `io::Result<()>` value without ignoring or unwrapping any such values.
* Changed the example and documentation to reflect the changes of the macro and to be a bit more concise, but hopefully still informative.

I had to do a bit of a rewrite of the macro in order to implement the above changes, but I believe the input syntax of the macro is otherwise unchanged.

I hope the result still works for you as expected. If not, please let me know. Thank you for the time and effort you put into the original issue and the design of this new API.

Cryptjar commented:
Looks very nice!
The syntax has finally converged to something reasonably convenient while flexible. And the solution for the Result propagation is also very fancy.

The only error I could spot is in the last line in the example of the macro doc. It states:
```rust
term_writeln!(term, [theme=theme] "Red, bold text" [reset])?;
```
But that line will be printed _green_ and bold.
